File: CommonCoreLegacy/src/main/java/com/skedgo/tripkit/common/model/RegionsResponse.java
package com.skedgo.tripkit.common.model;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Map;

import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;

public class RegionsResponse {
    @SerializedName("regions")
    private ArrayList<Region> regions;
    @SerializedName("modes")
    private Map<String, TransportMode> transportModeMap;

    @Nullable
    public ArrayList<Region> getRegions() {
        return regions;
    }

    @VisibleForTesting
    void setRegions(ArrayList<Region> regions) {
        this.regions = regions;
    }

    @Nullable
    public Collection<TransportMode> getTransportModes() {
        correctModeIds();
        return transportModeMap != null
            ? transportModeMap.values()
            : null;
    }

    @VisibleForTesting
    void setTransportModeMap(Map<String, TransportMode> transportModeMap) {
        this.transportModeMap = transportModeMap;
    }

    private void correctModeIds() {
        if (transportModeMap != null) {
            for (Map.Entry<String, TransportMode> entry : transportModeMap.entrySet()) {
                if (entry.getValue() != null) {
                    entry.getValue().setId(entry.getKey());
                }
            }
        }
    }
}

File: CommonCoreLegacy/src/main/java/com/skedgo/tripkit/common/model/region/RegionsResponse.kt
package com.skedgo.tripkit.common.model.region

import androidx.annotation.VisibleForTesting
import com.google.gson.annotations.SerializedName
import com.skedgo.tripkit.common.model.TransportMode

class RegionsResponse {
    @JvmField
    @set:VisibleForTesting
    @SerializedName("regions")
    var regions: ArrayList<Region>? = null

    @SerializedName("modes")
    private var transportModeMap: Map<String, TransportMode>? = null

    val transportModes: Collection<TransportMode>?
        get() {
            correctModeIds()
            return if (transportModeMap != null
            ) transportModeMap?.values
            else null
        }

    @VisibleForTesting
    fun setTransportModeMap(transportModeMap: Map<String, TransportMode>?) {
        this.transportModeMap = transportModeMap
    }

    private fun correctModeIds() {
        if (transportModeMap != null) {
            for ((key, value) in transportModeMap!!) {
                if (value != null) {
                    value.id = key
                }
            }
        }
    }
}